### ISSUE TITLE
Discussion: Bochscpu exit during context switch

### ISSUE THREAD

**Issue by x9090** [0vercl0k/wtf#38]
Bonjour!

First of all, as mentioned in the subject title, this is more like a discussion rather than a real issue itself as I couldn't find the Discussion tab under this repo :)

Here is the context:

Bochscpu exits deliberately when it detects context switching. This is a good approach as the fuzzer is not interested in capturing the code of other processes, especially you are targeting UM program which can reduce the performance overhead at the same time. However, it could be a limitation when your target is in KM. I have a scenario where the testcase buffer will be queued in kernel's work item, where the separate kernel thread will be invoked, and parse the testcase buffer. I believe this is a common scenario if we are targeting a more complex kernel module. With the current approach, stopping the CPU when context switching is detected, the fuzzer definitely could not reach the actual parsing functions.

Since I'm not familiar with the concept of TLB in-depth and I'm not sure what are the side effects of allowing context switching besides the performance overhead, so I'm checking with you does it make sense to allow CR3 change in _BochscpuBackend_t::TlbControlHook_ (i.e. let the bochscpu continue execution) if the user is meant to target kernel modules?

Thanks!

**Comment by 0vercl0k** [comment]
Bonjour 😊

Thanks for starting up this discussion, it's very interesting! I also tried to turn on the Discussions on this repository, let me know if it is accessible!

I think allowing `@cr3` changing is fine as long as you know what's gonna happen. In your case, it feels like you don't necessarily know when the work item or any asynchronous job will get picked up which means that anything can happen in between: code that you're not interested in accessing a file which would trigger an exit as `wtf` doesn't support devices.

In the scenario you describe, I wonder if you're harnessing your target at a too high level; what about snapshotting at the entry point of the function actually handling the work item? I think another way could also be to be fuzzing the same target with several snapshots to exercise different things in your target. Those are my random thoughts 😊

Also tagging @yrp604 as he might have some thoughts on the problem as well.

Cheers

**Comment by yrp604** [comment]
Axel hit the main point: more execution, less control over execution = higher chance of hitting a device access that _does_ end the testcase. There are a few other things to consider/flesh this answer out:

1. There is no hard problem with cr3 switching during a testcase. The tlb control name is inherited from bochs instrumentation hooks, as on a cr3 change traditionally the tlb is flushed -- thus the name. In some other projects that use bxcpu I've emulated past cr3 changes without issue in some circumstances.
2. One thing to keep in mind is that wtf/bxcpu has no interrupt controller, meaning there are no clock interrupts while a testcase is executing. This means the only type of context switch youll get is a cooperative one when a thread yields back to the scheduler. This makes the schedulers thread ordering very, very important. It also means things like, should the idle thread ever get scheduled you'll likely infloop. In this regard it is very much unlike normal execution.
3. Coverage also gets more complicated once you allow cr3 changes. You probably now want to record coverage indexed off `(cr3, rip)` instead of just `rip` -- wtf could be made to do this in the bxcpu case, but it would be more difficult under the virtualization backends.

Basically, to make this work the way you want in the general case, I think you'll need to puppet the scheduler. Under some snapshots it might Just Work which is probably worth trying.

**Comment by x9090** [comment]
Thanks for the helpful insights, @0vercl0k @yrp604! 

@0vercl0k I can see the Discussions tab now, thanks for your quick turnaround! To answer your question regarding my harness, I must admit that I'm a bit ambitious about harnessing the target at higher level function simply because there are some "pre-parse" routines that perform the sanity checks on the testcase which I think the mutated buffer should go through them since this will affect the normal execution flow eventually. I even thought about the suggestion you'd given - using and exercising multiple snapshots but there is give and take ultimately :)

@yrp604 Thanks for your explanation about the tlb's hook name. Anyway, I started to experiment with it by allowing cr3 switching, there are no immediate issues as far as my target's snapshot is concerned.
